### PR TITLE
[WIP] typescript definitions

### PR DESCRIPTION
**Pull request by PopGoesTheWza**
Issue #245 [Work In Progress]

Types are broken down into two files

- `specification.d.ts` attempts to follow strictly Stamp Specification v1.6
- `index.d.ts` exposes `stampit` extended implementation

For review:

- lookout for  special comments `// ?`, `// !` and `// TODO`

Sample code

```ts
import stampit, { compose, init } from 'stampit';

// default instance type (any)

const a = stampit().init(function(options) {
  const a = options.args[0];
  this.getA = () => {
    return a;
  };
});
a(); // Object -- so far so good.
a().getA(); // "a"

// typed instances

  interface CoordsInstance {
    x: number;
    y: number;
    coordsToString: () => string;
  }

  interface ColorInstance {
    color: string;
  }

  const Coords = compose<CoordsInstance>({
    initializers: [
      function({ x, y }) {
        this.x = x;
        this.y = y;
      },
    ],
    methods: {
      coordsToString() {
        return `(${this.x}, ${this.y})`;
      },
    },
  });
  const Point = compose<CoordsInstance & { toString: () => string }>(
    Coords,
    {
      methods: {
        toString() {
          return this.coordsToString();
        },
      },
    },
  );
  const Color = compose<ColorInstance>({
    initializers: [
      function({ color }) {
        this.color = color;
      },
    ],
  });
```

### CODE REVIEW

**Comment by koresar** [comment]
First of all, this is quite solid ground work! So much new learning for me.

I would leave few comments on the PR, but my TypeScript knowledge is limited (to C#). So, they will likely be irrelevant. :) Sorry!

**Comment by koresar** [comment]
For streamlining this PR we need to get in sync on stampit's terminology. I'm getting confused with the terminology used in this PR. Let me put it the way I understand:

* **Stamp** - is a factory function to create plain object **instances**. A **Stamp** always has property `.compose` which is a (copy of) **Compose function**.
* **Instance** - a plain JavaScript object created from a **Stamp**. (We usually do not use the word "instance" in any other context.)
* **Descriptor** - a JavaScript object (not always plain, can be a function) which contains **Stamp**'s metadata as properties
* **Compose function** - the function which creates **Stamps** from **composables**.
* **Composer function** - note the **R** (!) - the function which is kept in the `composers` metadata of a **descriptor**
* **Composable** - either a **Stamp** or a **Descriptor**

File: types/specification.d.ts [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+       * **WARNING** Avoid using two different `.init()` functions that expect different arguments.
+       * `.init()` functions that take arguments should not be considered safe to compose
+       * with other `.init()` functions that also take arguments. Taking arguments with
+       * an `.init()` function is an anti-pattern that should be avoided, when possible.
+       * @return A new object composed of the Stamps and prototypes provided.
+       */
+      (options?: PropertyMap, ...args: Array<any>): I;
+
+      /** Stamp metadata/composer function */
+      compose: ComposeMethod<I> & Descriptor<I>;
```
Comment (koresar) [inline types/specification.d.ts]: Oh! That is smart! :)

File: types/specification.d.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  type StampType<I> = I extends Factory<infer I> ? I
+                    : I extends Descriptor<infer I> ? I
+                    : I extends Instance ? I
+                    : never;
+
+  /** A composable object (`stamp` factory or Descriptor) */
+  type Composable<I, U = StampType<I>> = Factory<U> | Descriptor<U>;
+
+  /** Base `stamp` instance */
+  type Instance = { [s: string]: any; };
```
Comment (koresar) [inline types/specification.d.ts]: This comment got be confused. I'd change "stamp instance" to "object instance", or "object instance created from stamp".

File: types/specification.d.ts [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+  /**
+   * A factory function that will produce new objects using the
+   * prototypes that are passed in or composed.
```
Comment (koresar) [inline types/specification.d.ts]: "using the prototypes that are passed" ? Can I suggest rewording of the comment: "using the matadata that is attached to `.compose`"

File: types/specification.d.ts [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+   */
+  // TODO: interface Factory<I, F extends Factory<any> = Factory<I>>
+  interface Factory<I> {
```
Comment (koresar) [inline types/specification.d.ts]: I suggest renaming this to `Stamp` if possible. Or `StampFactory`. IDK

File: types/specification.d.ts [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+       * @return A new object composed of the Stamps and prototypes provided.
+       */
+      (options?: PropertyMap, ...args: Array<any>): I;
+
+      /** Stamp metadata/composer function */
+      compose: ComposeMethod<I> & Descriptor<I>;
+  }
+
+  /** Create a new stamp based on this descriptor */
+  interface ComposeMethod<I> {
```
Comment (koresar) [inline types/specification.d.ts]: Can I suggest a better name? This thing is used both - as a standalone function and as a method. Should we rename it to `ComposeFunction`?

File: types/specification.d.ts (hunk 3 above)
Comment (PopGoesTheWza) [reply types/specification.d.ts]: Blame the one who did the original `@types/stampit` package :)

I first tried to stick with the terms used in Stamp Specification and later used description from the `@types/stampit` package (which are often confusing IMHO)

Can you edit the files in this PR?

File: types/specification.d.ts (hunk 5 above)
Comment (PopGoesTheWza) [reply types/specification.d.ts]: in Specification it is referred to as the `.compose()`method.
I did not include the "Standalone `compose()` pure function (optional)" in `specification.d.ts`at this point.

File: types/specification.d.ts (hunk 4 above)
Comment (PopGoesTheWza) [reply types/specification.d.ts]: Let's do with `Stamp`

File: types/specification.d.ts (hunk 3 above)
Comment (koresar) [reply types/specification.d.ts]: Blame them! :)

I can't edit files in this PR since it's your repo. :)

However, how about I give you write access to the stampit repo?

File: types/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+        (...composables: Array<StampitComposable<I>>): Stamp<I>;
+    }
+
+    /** Any composable object (stamp or descriptor) */
+    // ! should have been StampitComposable
+    // type Composable<I> = Stamp<I> | Descriptor<I>;
+
+    // TODO: Add description
+    // ? shouldn't stampit.Option be stampit.(Extended)Descriptor really?
+    interface Options<I, F extends Specification.Factory<any> = Specification.Factory<I>> extends Omit<Specification.Descriptor<I, F>, 'initializers'> {
```
Comment (koresar) [inline types/index.d.ts]: Correct. A better name would be `ExtendedDescriptor`

Comment (koresar) [reply types/index.d.ts]: Keep in mind, that `ExtendedDescriptor` is only the "input" type. It is always converted to a regular `Descriptor` by stampit.

**Comment by PopGoesTheWza** [comment]
@koresar All descriptions in `specification.d.ts` have been reviewed and should now make more sense

File: types/specification.d.ts [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+   */
+  interface InitializerContext<I> {
+      /** The object instance being produced by the `Stamp`. If the initializer returns a value other than `undefined`, it replaces the instance. */
+      instance?: I;
+      /** A reference to the `Stamp` producing the instance. */
+      // TODO: investigate how to support extended Stamp interfaces
+      stamp?: Stamp<I>;
+      /** An array of the arguments passed into the `Stamp`, including the options argument. */
+      // ! above description from the specification is obscure
+      args?: Array<any>;
```
Comment (PopGoesTheWza) [inline types/specification.d.ts]: I am unsure if `args` is properly typed. The specification is not very clear here.

File: types/index.d.ts [hunk 8]
```diff
@@ -0,0 +1,414 @@
...
+        /** Initialization function(s) which will be called per each newly created instance. */
+        init?: Specification.Initializer<I> | Array<Specification.Initializer<I>>;
+        /** Initialization function(s) which will be called per each newly created instance. */
+        initializers?: Specification.Initializer<I> | Array<Specification.Initializer<I>>;
+        /** A configuration object to be shallowly assigned to Stamps */
+        conf?: Specification.PropertyMap & ThisType<F>;
+        /** A configuration object to be deeply merged to Stamps */
+        deepConf?: Specification.PropertyMap & ThisType<F>;
+        // TODO: Add description
+        name: string;
```
Comment (PopGoesTheWza) [inline types/index.d.ts]: Please propose a description

**Comment by PopGoesTheWza** [comment]
Closing this outdated PR
